api/handler.go

```diff
@@ -406,6 +386,80 @@ func (h *apiHandler) queryTotalUsage() http.HandlerFunc {
 	}
 }
 
+func (h *apiHandler) queryRealtimeViewership() http.HandlerFunc {
+	return func(rw http.ResponseWriter, r *http.Request) {
+		querySpec, httpErrorCode, errs := h.resolveRealtimeQuerySpec(r)
+		if len(errs) > 0 {
+			respondError(rw, httpErrorCode, errs...)
+			return
+		}
+		metrics, err := h.views.QueryRealtimeEvents(r.Context(), querySpec)
+		if err != nil {
+			respondError(rw, http.StatusInternalServerError, err)
+			return
+		}
+		respondJson(rw, http.StatusOK, metrics)
+	}
+}
+
+func (h *apiHandler) resolveQuerySpec(r *http.Request) (views.QuerySpec, int, []error) {
+	var (
+		from, err1 = parseInputTimestamp(r.URL.Query().Get("from"))
+		to, err2   = parseInputTimestamp(r.URL.Query().Get("to"))
+	)
+	if errs := nonNilErrs(err1, err2); len(errs) > 0 {
+		return views.QuerySpec{}, http.StatusBadRequest, errs
+	}
+
+	userId := callerUserId(r)
+	if userId == "" {
+		return views.QuerySpec{}, http.StatusInternalServerError, []error{errors.New("request not authenticated")}
+	}
+
+	qs := r.URL.Query()
+	assetID, streamID := qs.Get("assetId"), qs.Get("streamId")
+	spec := views.QuerySpec{
+		From:     from,
+		To:       to,
+		TimeStep: qs.Get("timeStep"),
+		Filter: views.QueryFilter{
+			UserID:     userId,
+			PlaybackID: qs.Get("playbackId"),
+			CreatorID:  qs.Get("creatorId"),
+		},
+		BreakdownBy: qs["breakdownBy[]"],
+	}
+
+	if err := h.views.Validate(spec, assetID, streamID); err != nil {
+		return views.QuerySpec{}, http.StatusInternalServerError, []error{err}
+	}
+
+	return spec, 0, []error{}
+}
+
+func (h *apiHandler) resolveRealtimeQuerySpec(r *http.Request) (views.QuerySpec, int, []error) {
+	spec, httpErrorCode, errs := h.resolveQuerySpec(r)
+	if spec.TimeStep != "" {
+		return views.QuerySpec{}, http.StatusBadRequest, []error{errors.New("timeStep is not supported for Realtime Viewership API")}
+	}
+	if spec.From == nil && spec.To != nil {
+		return views.QuerySpec{}, http.StatusBadRequest, []error{errors.New("param 'to' cannot be specified if 'from' is not defined")}
+	}
+	if spec.From != nil || spec.To != nil {
+		// If using time range, then we allow to query max 1 min before now(),
+		// because the current "per minute" aggregation may not be finalized yet
```

Comment:
It's still a bit unclear to me what it means to have a "real time" metric that is actually of a time range in the past. Could be simpler to not have this in the beginning? Or do we expect customers to need this, and if so, for what reason?

Reply:
We need it 🙃 

It's realtime, but time series data. Something you'd display in the chart. Like active viewers over the last 30 min. The first use-case is our Livpeer Studio Dashboard.